File: service/src/test/java/hu/bsstudio/robonaut/team/BroadcastingTeamServiceTest.java
package hu.bsstudio.robonaut.team;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.openMocks;

import hu.bsstudio.robonaut.team.model.DetailedTeam;
import hu.bsstudio.robonaut.team.model.Team;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

final class BroadcastingTeamServiceTest {

    private static final Team TEAM = Team.builder().build();
    private static final DetailedTeam DETAILED_TEAM_1 = DetailedTeam.builder().teamId(1).build();
    private static final DetailedTeam DETAILED_TEAM_2 = DetailedTeam.builder().teamId(2).build();
    private static final String TEAM_DATA_ROUTING_KEY = "team.teamData";

    private BroadcastingTeamService underTest;

    @Mock
    private RabbitTemplate mockTemplate;
    @Mock
    private TeamService mockService;

    @BeforeEach
    void setUp() {
        openMocks(this);
        this.underTest = new BroadcastingTeamService(mockTemplate, mockService);
    }

    @Test
    void shouldReturnDetailedTeamFromUnderLyingServiceAndSendItWhenTeamIsCreated() {
        when(mockService.addTeam(TEAM))
            .thenReturn(Mono.just(DETAILED_TEAM_1));

        final var result = underTest.addTeam(TEAM);

        StepVerifier.create(result)
            .expectNext(DETAILED_TEAM_1)
            .verifyComplete();
        verify(mockTemplate).convertAndSend(TEAM_DATA_ROUTING_KEY, DETAILED_TEAM_1);
    }

    @Test
    void shouldReturnDetailedTeamFromUnderLyingServiceAndSendItWhenTeamIsUpdated() {
        when(mockService.updateTeam(TEAM))
            .thenReturn(Mono.just(DETAILED_TEAM_1));

        final var result = underTest.updateTeam(TEAM);

        StepVerifier.create(result)
            .expectNext(DETAILED_TEAM_1)
            .verifyComplete();
        verify(mockTemplate).convertAndSend(TEAM_DATA_ROUTING_KEY, DETAILED_TEAM_1);
    }

    @Test
    void shouldReturnDetailedTeamFromUnderLyingServiceAndSendItWhenTeamIsUpdatedByAdmin() {
        when(mockService.updateTeam(DETAILED_TEAM_1))
            .thenReturn(Mono.just(DETAILED_TEAM_1));

        final var result = underTest.updateTeam(DETAILED_TEAM_1);

        StepVerifier.create(result)
            .expectNext(DETAILED_TEAM_1)
            .verifyComplete();
        verify(mockTemplate).convertAndSend(TEAM_DATA_ROUTING_KEY, DETAILED_TEAM_1);
    }

    @Test
    void shouldReturnDetailedTeamFromUnderLyingServiceAndSendItWhenFindingAllTeam() {
        when(mockService.findAllTeam())
            .thenReturn(Flux.just(DETAILED_TEAM_1, DETAILED_TEAM_2));

        final var result = underTest.findAllTeam();

        StepVerifier.create(result)
            .expectNext(DETAILED_TEAM_1)
            .expectNext(DETAILED_TEAM_2)
            .verifyComplete();
        verify(mockTemplate).convertAndSend(TEAM_DATA_ROUTING_KEY, DETAILED_TEAM_1);
        verify(mockTemplate).convertAndSend(TEAM_DATA_ROUTING_KEY, DETAILED_TEAM_2);
    }
}

File: service/src/test/kotlin/hu/bsstudio/robonaut/team/BroadcastingTeamServiceTest.kt
package hu.bsstudio.robonaut.team

import hu.bsstudio.robonaut.team.model.DetailedTeam
import hu.bsstudio.robonaut.team.model.Team
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.amqp.rabbit.core.RabbitTemplate
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.test.StepVerifier

internal class BroadcastingTeamServiceTest {

    @MockK
    private lateinit var mockTemplate: RabbitTemplate
    @MockK
    private lateinit var mockService: TeamService

    private lateinit var underTest: BroadcastingTeamService

    @BeforeEach
    internal fun setUp() {
        MockKAnnotations.init(this)
        underTest = BroadcastingTeamService(mockTemplate, mockService)
    }

    @Test
    internal fun `should return DetailedTeam from underlying service and send it when Team is created`() {
        every { mockService.addTeam(TEAM) } returns Mono.just(DETAILED_TEAM_1)
        every { mockTemplate.convertAndSend(TEAM_DATA_ROUTING_KEY, DETAILED_TEAM_1) } returns Unit

        val result = underTest.addTeam(TEAM)

        StepVerifier.create(result)
            .expectNext(DETAILED_TEAM_1)
            .verifyComplete()
    }

    @Test
    internal fun `should return DetailedTeam from underlying service and send it when Team is updated`() {
        every { mockService.updateTeam(TEAM) } returns Mono.just(DETAILED_TEAM_1)
        every { mockTemplate.convertAndSend(TEAM_DATA_ROUTING_KEY, DETAILED_TEAM_1) } returns Unit

        val result = underTest.updateTeam(TEAM)

        StepVerifier.create(result)
            .expectNext(DETAILED_TEAM_1)
            .verifyComplete()
    }

    @Test
    internal fun `should return DetailedTeam from underlying service and send it when Team is updated by admin`() {
        every { mockService.updateTeam(DETAILED_TEAM_1) } returns Mono.just(DETAILED_TEAM_1)
        every { mockTemplate.convertAndSend(TEAM_DATA_ROUTING_KEY, DETAILED_TEAM_1) } returns Unit

        val result = underTest.updateTeam(DETAILED_TEAM_1)

        StepVerifier.create(result)
            .expectNext(DETAILED_TEAM_1)
            .verifyComplete()
    }

    @Test
    internal fun `should return DetailedTeam from underlying service and send it when finding all Team`() {
        every { mockService.findAllTeam() } returns Flux.just(DETAILED_TEAM_1, DETAILED_TEAM_2)
        every { mockTemplate.convertAndSend(TEAM_DATA_ROUTING_KEY, DETAILED_TEAM_1) } returns Unit
        every { mockTemplate.convertAndSend(TEAM_DATA_ROUTING_KEY, DETAILED_TEAM_2) } returns Unit

        val result = underTest.findAllTeam()

        StepVerifier.create(result)
            .expectNext(DETAILED_TEAM_1)
            .expectNext(DETAILED_TEAM_2)
            .verifyComplete()
    }

    companion object {
        private val TEAM: Team = Team()
        private val DETAILED_TEAM_1: DetailedTeam = DetailedTeam(teamId = 1)
        private val DETAILED_TEAM_2: DetailedTeam = DetailedTeam(teamId = 2)
        private const val TEAM_DATA_ROUTING_KEY = "team.teamData"
    }
}
